// ng-in-viewport/index.es5.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/Subject'), require('@angular/common'), require('rxjs/Subscription')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', 'rxjs/Subject', '@angular/common', 'rxjs/Subscription'], factory) :
    (factory((global['ng-in-viewport'] = {}),global.ng.core,global.Subject,global.ng.common,global.Subscription));
}(this, (function (exports,core,Subject,common,Subscription) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var InViewportUtils;
    (function (InViewportUtils) {
        /**
         * @param {?} value
         * @return {?}
         */
        function isObject(value) {
            var /** @type {?} */ type = typeof value;
            return value != null && (type == 'object' || type == 'function');
        }
        InViewportUtils.isObject = isObject;
        /**
         * @param {?} value
         * @return {?}
         */
        function isObjectLiteral(value) {
            return Object.prototype.toString.call(value) === '[object Object]';
        }
        InViewportUtils.isObjectLiteral = isObjectLiteral;
    })(InViewportUtils || (InViewportUtils = {}));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var isObjectLiteral = InViewportUtils.isObjectLiteral;
    /** @enum {number} */
    var InViewportConfigDirection = {
        Both: 0,
        Vertical: 1,
        Horizontal: 2,
    };
    InViewportConfigDirection[InViewportConfigDirection.Both] = "Both";
    InViewportConfigDirection[InViewportConfigDirection.Vertical] = "Vertical";
    InViewportConfigDirection[InViewportConfigDirection.Horizontal] = "Horizontal";
    var InViewportConfig = /** @class */ (function () {
        function InViewportConfig(options) {
            var /** @type {?} */ exist = isObjectLiteral(options);
            this.rootElement = exist && options.hasOwnProperty('rootElement') ? options.rootElement : void 0;
            this.partial = exist && options.hasOwnProperty('partial') ? options.partial : true;
            this.direction = exist && options.hasOwnProperty('direction') ? options.direction : void 0;
        }
        Object.defineProperty(InViewportConfig.prototype, "rootElement", {
            get: /**
             * @return {?}
             */
            function () {
                return this._rootElement;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                this._rootElement = value && value.nodeType === 1 ? value : void 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(InViewportConfig.prototype, "partial", {
            get: /**
             * @return {?}
             */
            function () {
                return this._partial;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                this._partial = !!value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(InViewportConfig.prototype, "direction", {
            get: /**
             * @return {?}
             */
            function () {
                return this._direction;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                var /** @type {?} */ availableOptions = [
                    InViewportConfigDirection.Both,
                    InViewportConfigDirection.Vertical,
                    InViewportConfigDirection.Horizontal
                ];
                this._direction = availableOptions.indexOf(value) >= 0 ? value : InViewportConfigDirection.Both;
            },
            enumerable: true,
            configurable: true
        });
        return InViewportConfig;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var InViewportService = /** @class */ (function () {
        function InViewportService(ngZone) {
            this.ngZone = ngZone;
            this.registry = [];
            this.trigger$ = new Subject.Subject();
        }
        /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        InViewportService.prototype.addTarget = /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        function (target, rootElement) {
            var _this = this;
            this.ngZone.runOutsideAngular(function () { return _this.register(target, rootElement); });
        };
        /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        InViewportService.prototype.removeTarget = /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        function (target, rootElement) {
            var _this = this;
            this.ngZone.runOutsideAngular(function () { return _this.unregister(target, rootElement); });
        };
        /**
         * @param {?} element
         * @return {?}
         */
        InViewportService.prototype.getRootElement = /**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            return element && element.nodeType === 1 ? element : null;
        };
        /**
         * @param {?} rootElement
         * @return {?}
         */
        InViewportService.prototype.findRegistryEntry = /**
         * @param {?} rootElement
         * @return {?}
         */
        function (rootElement) {
            var _this = this;
            return this.registry.find(function (item) { return item.rootElement === _this.getRootElement(rootElement); });
        };
        /**
         * @param {?} entries
         * @return {?}
         */
        InViewportService.prototype.onChanges = /**
         * @param {?} entries
         * @return {?}
         */
        function (entries) {
            var _this = this;
            if (Array.isArray(entries) && entries.length) {
                entries.forEach(function (entry) { return _this.trigger$.next(entry); });
            }
        };
        /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        InViewportService.prototype.register = /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        function (target, rootElement) {
            var _this = this;
            var /** @type {?} */ registryEntry = this.findRegistryEntry(rootElement);
            if (!registryEntry) {
                var /** @type {?} */ registryEntryObserverOptions = {
                    root: this.getRootElement(rootElement),
                    threshold: Array(101)
                        .fill(void 0)
                        .map(function (item, i) { return i / 100; })
                };
                registryEntry = {
                    targets: [target],
                    rootElement: this.getRootElement(rootElement),
                    observer: new IntersectionObserver(function (entries) { return _this.ngZone.run(function () { return _this.onChanges(entries); }); }, registryEntryObserverOptions)
                };
                registryEntry.observer.observe(target);
                this.registry.push(registryEntry);
            }
            else if (registryEntry.targets.indexOf(target) < 0) {
                registryEntry.targets.push(target);
                registryEntry.observer.observe(target);
            }
        };
        /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        InViewportService.prototype.unregister = /**
         * @param {?} target
         * @param {?=} rootElement
         * @return {?}
         */
        function (target, rootElement) {
            var /** @type {?} */ registryEntry = this.findRegistryEntry(rootElement);
            var /** @type {?} */ registryEntryIdx = this.registry.indexOf(registryEntry);
            if (registryEntry) {
                var /** @type {?} */ targetIdx = registryEntry.targets.indexOf(target);
                if (targetIdx >= 0) {
                    registryEntry.observer.unobserve(target);
                    registryEntry.targets.splice(targetIdx, 1);
                }
                if (registryEntry.targets.length === 0) {
                    registryEntry.observer.disconnect();
                    this.registry.splice(registryEntryIdx, 1);
                }
            }
        };
        InViewportService.decorators = [
            { type: core.Injectable },
        ];
        /** @nocollapse */
        InViewportService.ctorParameters = function () { return [
            { type: core.NgZone, },
        ]; };
        return InViewportService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var isObjectLiteral$1 = InViewportUtils.isObjectLiteral;
    var /** @type {?} */ InViewportMetadata = Symbol('InViewportMetadata');
    var InViewportDirective = /** @class */ (function () {
        function InViewportDirective(platformId, elementRef, inViewportService) {
            this.platformId = platformId;
            this.elementRef = elementRef;
            this.inViewportService = inViewportService;
            this.subscription = new Subscription.Subscription();
            this.config = new InViewportConfig();
            this.action$ = new core.EventEmitter();
        }
        Object.defineProperty(InViewportDirective.prototype, "updateConfig", {
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (isObjectLiteral$1(value)) {
                    if (value.hasOwnProperty('rootElement')) {
                        this.config.rootElement = value.rootElement;
                    }
                    if (value.hasOwnProperty('partial')) {
                        this.config.partial = value.partial;
                    }
                    if (value.hasOwnProperty('direction')) {
                        this.config.direction = value.direction;
                    }
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        InViewportDirective.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (common.isPlatformBrowser(this.platformId)) {
                this.subscription.add(this.inViewportService.trigger$.subscribe(function (entry) { return _this.check(entry); }));
                this.inViewportService.addTarget(this.elementRef.nativeElement, this.config.rootElement);
            }
            else {
                this.check(undefined, true);
            }
        };
        /**
         * @return {?}
         */
        InViewportDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.subscription.unsubscribe();
            if (common.isPlatformBrowser(this.platformId)) {
                this.inViewportService.removeTarget(this.elementRef.nativeElement, this.config.rootElement);
            }
        };
        /**
         * @param {?} entry
         * @param {?=} force
         * @return {?}
         */
        InViewportDirective.prototype.check = /**
         * @param {?} entry
         * @param {?=} force
         * @return {?}
         */
        function (entry, force) {
            if (force || (entry && entry.target === this.elementRef.nativeElement)) {
                var /** @type {?} */ value = force || (this.config.partial ? entry.intersectionRatio > 0 : entry.intersectionRatio === 1);
                this.action$.emit((_a = {},
                    _a[InViewportMetadata] = { entry: entry },
                    _a.target = this.elementRef.nativeElement,
                    _a.value = value,
                    _a));
            }
            var _a;
        };
        InViewportDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[inViewport], [in-viewport]'
                    },] },
        ];
        /** @nocollapse */
        InViewportDirective.ctorParameters = function () { return [
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] },] },
            { type: core.ElementRef, },
            { type: InViewportService, },
        ]; };
        InViewportDirective.propDecorators = {
            "action$": [{ type: core.Output, args: ['inViewportAction',] },],
            "updateConfig": [{ type: core.Input, args: ['inViewportOptions',] },],
        };
        return InViewportDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var InViewportModule = /** @class */ (function () {
        function InViewportModule() {
        }
        /**
         * @return {?}
         */
        InViewportModule.forRoot = /**
         * @return {?}
         */
        function () {
            return {
                ngModule: InViewportModule,
                providers: [InViewportService]
            };
        };
        InViewportModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [InViewportDirective],
                        exports: [InViewportDirective]
                    },] },
        ];
        return InViewportModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.InViewportModule = InViewportModule;
    exports.InViewportService = InViewportService;
    exports.InViewportConfig = InViewportConfig;
    exports.InViewportMetadata = InViewportMetadata;
    exports.InViewportConfigDirection = InViewportConfigDirection;
    exports.InViewportDirective = InViewportDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
